### PR TITLE
Add AVX512 sort feature during creating index

### PR DESCRIPTION
**Pull request by xtangxtang**
### What changes were proposed in this pull request?
This change is to add avx512 sort during create Index. Before this change std::sort is used during creating index. Avx512 sort is intruduced by paper "A Novel Hybrid Quicksort Algorithm Vectorized using AVX-512 on Intel Skylake". 

### Why are the changes needed?
After adopting the change. creating index can speed up from 20% ~ 47% according to different index type. The performance workload is SSB. The table and creating index steps is as following SQL:
    
    create table lineitem (
    l_orderkey bigint not null,
    l_partkey bigint not null,
    l_suppkey bigint not null,
    l_linenumber bigint not null,
    l_quantity bigint not null,
    l_extendedprice bigint not null,
    l_discount bigint not null,
    l_tax bigint not null,
    l_returnflag char(1) default null,
    l_linestatus char(1) default null,
    l_shipdate date not null,
    l_commitdate date default null,
    l_receiptdate date default null,
    l_shipinstruct char(25) default null,
    l_shipmode char(10) default null,
    l_comment varchar(44) default null,
    primary key(l_orderkey, l_linenumber))
    tablegroup = tpch_tg_100g_lineitem_order_group
    partition by key (l_orderkey) partitions 192;
    
    create index I_L_ORDERKEY on lineitem(l_orderkey) local;
    create index I_L_SHIPINS on lineitem(l_shipinstruct) local;
    create index I_L_SHIPDATE_COMMITDATE on lineitem(l_shipdate, l_commitdate) local;
    
    drop index I_L_ORDERKEY on lineitem;
    drop index I_L_SHIPINS on lineitem;
    drop index I_L_SHIPDATE_COMMITDATE on lineitem;    

The performance is test on Intel(R) Xeon(R) Gold 6252 CPU @ 2.10GHz and the results is (unit is second):    

image

### Will break the compatibility? How if so?

Because the AVX512 is only available on Intel CPU and even since Intel Skylake CPU can support AVX512.  Two conditions are added to avoid breaking the compability:

* ENABLE_AVX512F is a new enviroment in build.sh. The OceanBase will compile the AVX512 code in case  ENABLE_AVX512F = ON
* __builtin_cpu_supports("avx512f")：is  added in src/storage/ob_parallel_external_sort.h. It will runtime check if current CPU surpport AVX512.

So these two switcheds decide if the AVX512 sort is running. 

image

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
The patch currently does not run the unit test.

### Checklist

- [X ] I've run the tests to see all new and existing tests pass.
- [X ] If this Pull Request resolves an issue, I linked to the issue in the text above.
- [X ] I've informed the technical writer about the documentation change if necessary.

### Intruduction for this change
The attached PDF summary the changes for this path
[OSD_AVX512_Eanbling_README.MD at main · epeshared_OSD_AVX512_Eanbling.pdf](https://github.com/oceanbase/oceanbase/files/8549078/OSD_AVX512_Eanbling_README.MD.at.main.epeshared_OSD_AVX512_Eanbling.pdf)

### CODE REVIEW

File: unittest/storage/test_parallel_external_sort.cpp [hunk 1]
```diff
@@ -57,13 +57,28 @@ class TestItem {
...
+    char* ptr = reinterpret_cast<char*>(&key_);
+    memcpy(key_buf, ptr, 8);
+    next_ids_++;
+    return oceanbase::common::OB_SUCCESS;
   }
+  bool get_sort_key_end() 
+  {
+    return next_ids_ >= 1;
+  }  
+
```
Comment (hnwyllmm) [inline unittest/storage/test_parallel_external_sort.cpp:74]: 这段代码是因为avx512比较是按照二进制比较的，所以需要加个函数get_sort_key来帮助将任意内型的key转换成可以按照二进制比较的key。因为TestItem里面的key都是int64的，所以可以直接通过内存拷贝将int64转成可以利用avx512比较的二进制形式。
对于加AVX宏开关，因为如果不走avx512分支的话，get_sort_key是不会被调用的，所以当avx没打开的话，这个函数是不会被调用的。但是也尅加上开个，这样编译的时候就不会编译这个函数的了。

**Review by hnwyllmm — APPROVED** [review]
LGTM

File: run/observer.pid [hunk 2]
```diff
@@ -0,0 +1 @@
+3514848
```
Comment (hnwyllmm) [inline run/observer.pid:1]: 测试代码不需要提交上来。

File: deps/oblib/src/common/object/ob_object.cpp [hunk 3]
```diff
@@ -512,6 +519,498 @@ int ObObj::build_not_strict_default_value()
...
+        if (copied < size) {
+          if (str_offset < val_len) {
+            int32_t need = size - copied;
+            int32_t t1 = (offset - 1) % 9;
+            int32_t t2 = ((offset - 1) + need) % 9;
+            int32_t sortkey_start = (offset - 1) / 9 * 8 + t1;
+            int32_t sortkey_end = ((offset - 1) + need) / 9 * 8 + t2;
+            int64_t sortkey_len = sortkey_end - sortkey_start;
+            int64_t str_len = val_len - str_offset;
+            bool is_valid_collation = false;
```
Comment (hnwyllmm) [inline deps/oblib/src/common/object/ob_object.cpp:870]: 可以解析下这段代码吗？谢谢

File: deps/oblib/src/common/object/ob_object.cpp [hunk 4]
```diff
@@ -512,6 +519,498 @@ int ObObj::build_not_strict_default_value()
...
+ * example:
+ * ob_object's data(v_) -> (char* array)[ null-flag | b0 | b1 | b2 | b3 | b4 | b5 | b6 | b7 | b8 | ...]
+ *
+ * to          : buffer to store the converted byte array
+ * offset      : represents the offset of this byte array
+ * size        : represents the length of the buffer(to)
+ * extra_param : help convert multi-byte encoded data to mencomparable format
+ *
+ */
+int ObObj::make_sort_key(char* to, int16_t& offset, int32_t& size,
```
Comment (hnwyllmm) [inline deps/oblib/src/common/object/ob_object.cpp:540]: 需要增加此函数的单元测试

File: deps/oblib/src/common/object/ob_object.cpp [hunk 5]
```diff
@@ -512,6 +519,498 @@ int ObObj::build_not_strict_default_value()
...
+            copied++;
+            offset++;
+          }
+        }
+        size = copied;
+      }
+      break;
+    }
+    case ObNumberType: 
+    case ObUNumberType: {
```
Comment (hnwyllmm) [inline deps/oblib/src/common/object/ob_object.cpp:948]: 类型少了一些，需要覆盖全量的，比如hexstring，text等等

File: src/storage/ob_avx512_sort.h [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+namespace oceanbase {
+namespace storage {
+
+
+int64_t lg(int64_t n);
+void do_sort(uint64_t* keys, uint64_t* values, const int64_t left, 
+               const int64_t right, int64_t depth_limit);
+
+template <typename T>
+class SortWithAvx512 {
```
Comment (hnwyllmm) [inline src/storage/ob_avx512_sort.h:34]: 整个算法的单测需要补下的。

**Review by hnwyllmm — CHANGES_REQUESTED** [review]
需要做些针对性的调整。

File: run/observer.pid (hunk 2 above)
Comment (xtangxtang) [reply run/observer.pid]: 在新的commit中已经删除

File: deps/oblib/src/common/object/ob_object.cpp (hunk 4 above)
Comment (xtangxtang) [reply deps/oblib/src/common/object/ob_object.cpp:540]: 在新的commit
https://github.com/oceanbase/oceanbase/pull/887/commits/a1becfd40f6ad717ec390f20db9964f1db8df0a5
中对test_parallel_external_sort.cpp新增了UT test_avx512_sort_type

File: deps/oblib/src/common/object/ob_object.cpp (hunk 5 above)
Comment (xtangxtang) [reply deps/oblib/src/common/object/ob_object.cpp:948]: 在新的commit中对make_sort_key做了如果遇到不支持的类型，将返回OB_NOT_SUPPORTED错误码，然后在ob_parallel_external_sort.h的finish函数和build_fragment函数中判断如果返回值是OB_NOT_SUPPORTED，将排序算法切换到以前的std::sort。

File: deps/oblib/src/common/object/ob_object.cpp (hunk 4 above)
Comment (xtangxtang) [reply deps/oblib/src/common/object/ob_object.cpp:540]: 是否需要对make_sort_key所有支持的type做验证？

File: src/storage/ob_avx512_sort.h (hunk 6 above)
Comment (xtangxtang) [reply src/storage/ob_avx512_sort.h:34]: 在https://github.com/oceanbase/oceanbase/pull/887/commits/1893e1ec369fa1f35f7e309ca759da8582757596
这个commit中加了SortWithAvx512.sort的UTcase

File: deps/oblib/src/common/object/ob_object.cpp (hunk 4 above)
Comment (xtangxtang) [reply deps/oblib/src/common/object/ob_object.cpp:540]: 在 https://github.com/oceanbase/oceanbase/pull/887/commits/d44db7097397558ffb108aeedd349cb112b70cf6 这个commit中新加了test_avx512_sort_type UT对全部ob type做了验证。

File: deps/oblib/src/common/object/ob_object.cpp (hunk 5 above)
Comment (xtangxtang) [reply deps/oblib/src/common/object/ob_object.cpp:948]: 在 https://github.com/oceanbase/oceanbase/commit/d44db7097397558ffb108aeedd349cb112b70cf6 这个commit中新加了test_avx512_sort_type UT对全部ob type做了验证。

File: src/storage/ob_avx512_sort.h (hunk 6 above)
Comment (xtangxtang) [reply src/storage/ob_avx512_sort.h:34]: 在 https://github.com/oceanbase/oceanbase/commit/d44db7097397558ffb108aeedd349cb112b70cf6 这个commit中新加了test_avx512_sort_type UT对SortWithAvx512.sort做了验证

File: deps/oblib/src/common/object/ob_object.cpp (hunk 3 above)
Comment (xtangxtang) [reply deps/oblib/src/common/object/ob_object.cpp:870]: 这段代码前面用t1表示在一个9个字节的数据块中的需要拷贝的起始位置，t2表示在一个9个字节的数据块中拷贝的结束位置。因为t1和t2之间可能会跨越多个数据块。所以需要用sortkey_start表示需要拷贝的起始数据位置，sortkey_end表示需要拷贝的结束数据位置。（(offset - 1) / 9 * 8表示需要拷贝的起始数据块，((offset - 1) + need) / 9 * 8表示需要拷贝的结束数据块位置）

**Comment by LINxiansheng** [comment]
transfer #1325

**Pull request closed without merging**